finetuner/tuner/dataset/__init__.py

```diff
@@ -1,8 +1,175 @@
 import itertools
+from typing import Any, Callable, Dict, Generic, List, Optional, Tuple, TypeVar, Union
 
 import numpy as np
 
 from ... import __default_tag_key__
+from ...helper import DocumentSequence
+
+
+AnyLabel = TypeVar('AnyLabel')
+
+
+class BaseDataset(abc.ABC, Generic[AnyLabel]):
+    _labels: List[AnyLabel]
+
+    @abc.abstractmethod
+    def __getitem__(self, ind: int) -> Tuple[Union[np.ndarray, str], AnyLabel]:
+        """
+        Get the (preprocessed) content and label for the item at ``ind`` index in the
+        dataset.
+        """
+
+    @property
+    def labels(self) -> List[AnyLabel]:
+        """ Get the list of labels for all items in the dataset."""
+        return self._labels
+
+    def __len__(self) -> int:
+        return len(self._labels)
+
+
+class ClassDataset(BaseDataset[int]):
+    """ Dataset for enapsulating data where each item has a class label."""
+
+    def __init__(
+        self,
+        docs: DocumentSequence,
+        preprocess_fn: Optional[Callable[[Union[str, np.ndarray]], Any]] = None,
+    ):
+        """Create the :class:`ClassDataset` instance.
+
+        :param docs: The documents for the dataset. Each document is expected to have
+            - a content (only blob or text are accepted currently)
+            - a class label, saved under ``tags['finetuner']['label']``. This class
+              label should be an integer or a string
+        :param preprocess_fn: A pre-processing function. It should take as input the
+            content of an item in the dataset (currently only text of blob are
+            accepted).
+        """
+        self._docs = docs
+        self._preprocess_fn = preprocess_fn
+
+        self._tag_labels_dict: Dict[Union[str, int], int] = {}
+        self._labels: List[int] = []
+        max_label = 0
+
+        for doc in self._docs:
+            try:
+                tag = doc.tags[__default_tag_key__]['label']
+            except KeyError as e:
+                raise KeyError(
+                    'The tag ["finetuner"]["label"] was not found in a document.'
+                    ' When using ClassDataset all documents need this tag'
+                ) from e
+            label = self._tag_labels_dict.get(tag)
+
+            if label is None:
+                label = max_label
+                self._tag_labels_dict[tag] = label
+                max_label += 1
+
+            self._labels.append(label)
+
+    def __getitem__(self, ind: int) -> Tuple[Union[np.ndarray, str], int]:
+        """
+        Get the (preprocessed) content and label for the item at ``ind`` index in the
+        dataset.
+        """
+        content = self._docs[ind].content
+        label = self._labels[ind]
+
+        if self._preprocess_fn:
+            content = self._preprocess_fn(content)
+
+        return (content, label)
+
+    @property
+    def labels(self) -> List[int]:
+        """ Get the list of integer labels for all items in the dataset."""
+        return self._labels
```

Comment:
why implement the same property again in child class? seems the type hints in base class and child class is different as well?

Reply:
Regarding different type hints: this is exactly how generics work, see this for example: https://stackoverflow.com/questions/57521599/python-generics-and-subclasses

Same property is implemented again because the docstring is different.